### PR TITLE
Add more text related tutorials

### PR DESCRIPTION
**Pull request by zechengz**

### CODE REVIEW

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 1]
```diff
@@ -1,38 +1,54 @@
...
-which use a text embedding model to pre-encode text columns into embeddings
-(frozen during model training). We plan to support fine-tuning of the text embedding
-models soon.
+:class:`stype.text_embedded<torch_frame.stype>` and :class:`stype.text_tokenized<torch_frame.stype>`
+for text columns.
+:class:`stype.text_embedded<torch_frame.stype>` uses text embedding model(s) to pre-encode text columns into embeddings
+(those models are frozen during model training).
+:class:`stype.text_tokenized<torch_frame.stype>` finetunes underlying text embedding models by using the
+gradients backpropagated from the tabular learning. Now :pyf:`PyTorch Frame` pipeline supports supervised finetuning
+on text models with other libraries such as the `PEFT <https://huggingface.co/docs/peft/>`_.
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:12]: "Now ..." sounds a bit awkward to me. Can you rewrite to just say `text_tokenized` supports finetuning with peft?

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 2]
```diff
@@ -1,38 +1,54 @@
...
 with text columns, such as :obj:`~torch_frame.datasets.MultimodalTextBenchmark`.
 
 In :pyf:`PyTorch Frame`, you can specify text columns as
 :class:`stype.text_embedded<torch_frame.stype>`. This will
-encode text columns using a user-specified text embedding model during the
+encode text columns using a user-specified text embedding model(s) during the
 dataset materialization stage.
+You can also specify text columns as
+:class:`stype.text_tokenized<torch_frame.stype>`, which will finetune
+user-specified text model(s) during the training stage.
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:30]: you can specify the stype of text columns in two ways.
begin{itemize}
1. stype.text_embedded: explain
2. stype.text_tokenized: explain

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 3]
```diff
@@ -1,38 +1,54 @@
...
 The processes of initializing and materializing datasets are similar
 to :doc:`/get_started/introduction`. Below we highlight the difference.
 
-First you need to specify your text embedding model. Here, we use the
+
+Using Pre-trained Text Embeddings
+---------------------------------
+
+For text columns with :class:`stype.text_embedded<torch_frame.stype>` that
+utilizes pre-trained text embeddings,
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:40]: shorter sentence will read better

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 4]
```diff
@@ -1,38 +1,54 @@
...
+You can also specify text columns as
+:class:`stype.text_tokenized<torch_frame.stype>`, which will finetune
+user-specified text model(s) during the training stage.
 
 The processes of initializing and materializing datasets are similar
 to :doc:`/get_started/introduction`. Below we highlight the difference.
 
-First you need to specify your text embedding model. Here, we use the
+
+Using Pre-trained Text Embeddings
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:36]: Add stype.text_embedded

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 5]
```diff
@@ -62,23 +77,27 @@ Then we instantiate :obj:`~torch_frame.config.TextEmbedderConfig` for our text e
...
-    text_embedder_cfg = TextEmbedderConfig(text_embedder=text_encoder,
-                                       batch_size=5)
+    text_embedder = TextToEmbedding(device)
+    col_to_text_embedder_cfg = TextEmbedderConfig(text_embedder=text_embedder, batch_size=5)
 
 Here :obj:`text_embedder` maps a list of sentences into PyTorch Tensor embeddings
 in mini-batch, where :obj:`batch_size` represents the batch size.
+Also, notice that we allow user to specify a dictionary of :obj:`text_embedder`s for different
+text columns with :class:`stype.text_embedded<torch_frame.stype>`. This also allows using
+:obj:`text_embedder`s with different embedding size for different text columns.
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:87]: this allows (delete "also")

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 6]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
 Then, :obj:`stype_encoder_dict` can be directly fed into
 :class:`~torch_frame.nn.encoder.StypeWiseFeatureEncoder` to handle text columns.
 
+
+Finetuning Underlying Text Models
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:160]: mention in the title that this is about `text_tokenized`

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 7]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
...
+---------------------------------
+
+For text columns with :class:`stype.text_tokenized<torch_frame.stype>`
+that finetunes underlying text models during the tabular learning, you need to specify both
+of the tokenization and encoding.
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:165]: both tokenizer and text embedding model that embeds tokenized sentences.

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 8]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
 Then, :obj:`stype_encoder_dict` can be directly fed into
 :class:`~torch_frame.nn.encoder.StypeWiseFeatureEncoder` to handle text columns.
 
+
+Finetuning Underlying Text Models
+---------------------------------
+
+For text columns with :class:`stype.text_tokenized<torch_frame.stype>`
+that finetunes underlying text models during the tabular learning, you need to specify both
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:164]: I think we first need to establish that many deep learning NLP models follow two-stage pipeline. first tokenize (mapping string into integers) and then embed the tokenized sentences.

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 9]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
...
+Here, we use the
+`Transformers <https://huggingface.co/docs/transformers>`_ package.
+
+.. code-block:: bash
+
+    pip install transformers
+
+
+Specifying Text Tokenization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:175]: shouldn't this subsection?

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 10]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
...
+    # A dictionary of text tokenization results
+    dataset.tensor_frame.feat_dict[torch_frame.text_tokenized]
+    >>> {'input_ids': MultiNestedTensor(num_rows=105154, num_cols=1, device='cpu'), 'attention_mask': MultiNestedTensor(num_rows=105154, num_cols=1, device='cpu')}
+
+
+Finetuning Text Models with Tabular Learning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To finetune the underlying text models together with tabular learning,
+we need at first specify how to embed tokenization results to text embeddings and how to finetune the text model.
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:250]: polish sentence

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 11]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
...
+.. code-block:: python
+
+    from torch_frame.config.text_tokenizer import TextTokenizerConfig
+
+    text_tokenizer = TextToEmbeddingTokenization()
+    col_to_text_tokenizer_cfg = TextTokenizerConfig(text_tokenizer=text_tokenizer, batch_size=10000)
+
+
+Here :obj:`text_tokenizer` maps a list of sentences into a dictionary of PyTorch Tensors,
+in mini-batch, where :obj:`batch_size` represents the batch size.
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:209]: polish sentence

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 12]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
...
+            torch_frame.text_tokenized]
+    }
+
+    stype_encoder_dict = {
+        stype.categorical: EmbeddingEncoder(),
+        stype.numerical: LinearEncoder(),
+        stype.text_tokenized: LinearModelEncoder(col_to_model_cfg=col_to_model_cfg)
+    }
+
+We provides :class:`~torch_frame.config.ModelConfig` to specify the text model to finetune and its output size.
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:336]: We provide

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 13]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
...
+
+    dataset.col_to_stype['description']
+    >>> <stype.text_tokenized: 'text_tokenized'>
+
+    # Materialize will call tokenizer for text columns
+    dataset.materialize()
+
+    # A dictionary of text tokenization results
+    dataset.tensor_frame.feat_dict[torch_frame.text_tokenized]
+    >>> {'input_ids': MultiNestedTensor(num_rows=105154, num_cols=1, device='cpu'), 'attention_mask': MultiNestedTensor(num_rows=105154, num_cols=1, device='cpu')}
```
Comment (weihua916) [inline docs/source/handling_advanced_stypes/handle_text.rst:243]: Can we briefly explain what `MultiNestedTensor` is and why we use it here?

**Review by weihua916 — COMMENTED** [review]
Thanks! Left some initial feedback.

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 14]
```diff
@@ -62,23 +77,27 @@ Then we instantiate :obj:`~torch_frame.config.TextEmbedderConfig` for our text e
...
+    col_to_text_embedder_cfg = TextEmbedderConfig(text_embedder=text_embedder, batch_size=5)
 
 Here :obj:`text_embedder` maps a list of sentences into PyTorch Tensor embeddings
 in mini-batch, where :obj:`batch_size` represents the batch size.
+Also, notice that we allow user to specify a dictionary of :obj:`text_embedder`s for different
+text columns with :class:`stype.text_embedded<torch_frame.stype>`. This also allows using
+:obj:`text_embedder`s with different embedding size for different text columns.
+
+Embedding Text Columns for a Dataset
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (yiweny) [inline docs/source/handling_advanced_stypes/handle_text.rst:90]: Add some transition sentence here?

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 15]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
...
+.. code-block:: bash
+
+    pip install transformers
+
+
+Specifying Text Tokenization
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Different from text columns with :class:`stype.text_embedded<torch_frame.stype>`, text columns with
+:class:`stype.text_tokenized<torch_frame.stype>` will be tokenized at first during the materialization
```
Comment (yiweny) [inline docs/source/handling_advanced_stypes/handle_text.rst:178]:
```suggestion
:class:`stype.text_tokenized` columns will be tokenized at first during materialization.
```

File: docs/source/handling_advanced_stypes/handle_text.rst [hunk 16]
```diff
@@ -132,6 +156,197 @@ As mentioned earlier, :class:`~torch_frame.stype.text_embedded` is stored togeth
...
+
+    from torch_frame.config.text_tokenizer import TextTokenizerConfig
+
+    text_tokenizer = TextToEmbeddingTokenization()
+    col_to_text_tokenizer_cfg = TextTokenizerConfig(text_tokenizer=text_tokenizer, batch_size=10000)
+
+
+Here :obj:`text_tokenizer` maps a list of sentences into a dictionary of PyTorch Tensors,
+in mini-batch, where :obj:`batch_size` represents the batch size.
+Also, notice that we allow user to specify a dictionary of :obj:`text_tokenizer`s for different
```
Comment (yiweny) [inline docs/source/handling_advanced_stypes/handle_text.rst:210]: This sentence doesn't render properly in docs.

**Review by yiweny — APPROVED** [review]

File: docs/source/handling_advanced_stypes/handle_text.rst (hunk 9 above)
Comment (zechengz) [reply docs/source/handling_advanced_stypes/handle_text.rst:175]: This is a subsubsection which lies in the subsection
```
Using Pre-trained Text Embeddings
---------------------------------
```

**Review by weihua916 — APPROVED** [review]
Thanks. @zechengz I made many edits. Can you please double-check them and feel free to merge.

**Pull request merged**